### ISSUE TITLE
Provide option to generate UUID string representation that will round-trip between .NET and Java/Javascript

### ISSUE THREAD

**Issue by fourpastmidnight** [abock/idgen#3]
First, very nice tool!

Second, it follows the UUID spec very well. But there is one issue: the generated UUID string cannot be used as an argument to `System.Guid(string)` constructor which would result in an identical Guid. In other words this is currently not true:

```csharp
Console.WriteLine(
    new Guid("de9425a4-e8dd-510b-8e00-b6ac890c733a") == new Guid(
        new byte[] {
            0xde, 0x94, 0x25, 0xa4,
            0xe8, 0xdd,
            0x51, 0x0b,
            0x8e, 0x00, 0xb6, 0xac, 0x89, 0x0c, 0x73, 0x3a
        }
    )
);

// false
```

The `System.Guid` implementation stores Guids internally in the following format:

```csharp
public struct Guid
{
    public int a;
    public short b;
    public short c;
    public byte d;
    public byte e;
    public byte f;
    public byte g;
    public byte h;
    public byte i;
    public byte j;
    public byte k;
}
```

In .NET, integers are stored in machine-byte order. And for Intel-based machines, that's Little-Endian. (Internally, `Guid` will convert a given byte array to store the bytes in the fields shown above. Oddly enough, when calling `ToByteArray()` on a `Guid`, it "undoes" the conversion and outputs the bytes in network byte order (i.e. Big-Endian). And even more oddly, `.ToString()` simply shows the string representation of the individual components, concatenated in a format complying with how UUIDs are segmented according to RFC4122, _BUT_, no conversion of (specifically) the `a`, `b` and `c` components is performed. So they display in their Little-Endian format. Way to go MS for standards-based compliance and interoperability.

I appreciate that this tool follows the UUID spec and provides for similarly formatted output as to what you'd get by calling `Guid.ToString`. However, it would be nice to provide an option to have the tool output its result that can be used directly with the `System.Guid(string)` constructor with no further conversion and that would result in the Guid's internal byte representation resulting in the exact same UUID. In other words, it would nice to be able to do this:

```powershell
$g = [Guid]::new($(dotnet idgen v5 -le bojangles '11de2b26-984e-56b4-aa25-b3bd28ea5ac2'))
$areEqual = (
    "$($g.ToByteArray() | % { $_.ToString('x2') })" -replace ' '
) -eq (
    $(dotnet idgen v5 bojangles '11de2b26-984e-56b4-aa25-b3bd28ea5ac2') -replace '-'
)

$areEqual
# True
```

The reason this is important is for interoperability between, say, Java (and Javascript) code and .NET code.

So using  your documentation's `bojangles` example to further demonstrate the issue:

```csharp
var g1 = new Guid("11de2b26-984e-56b4-aa25-b3bd28ea5ac2");
// Now hash 'bojangles' and use the UUID v5 algorithm to create a new GUID
var namespaceBytes = g1.ToByteArray();
var bojanglesBytes = System.Text.Encoding.UTF8.GetBytes("bojangles");

// Create a new byte array to hold the new GUID bytes
var bytes = new byte[namespaceBytes.Length + bojanglesBytes.Length];

Array.Copy(namespaceBytes, bytes, namespaceBytes.Length);
Array.Copy(bojanglesBytes, 0, bytes, namespaceBytes.Length, bojanglesBytes.Length);

// Now SHA1 hash the resulting byte array
var sha1Hasher = System.Security.Cryptography.SHA1CryptoServiceProvider.Create();
var bojanglesUUIDHash = sha1Hasher.ComputHash(bytes);

// We're creating a v5 UUID...
bojanglesUUIDHash[6] = ((bojanglesUUIDHash[6] & 0x0F) | 0x50);

// And conform to RFC4122...
bojanglesUUIDHash[8] = ((bojanglesUUIDHash[8] & 0x3f) | 0x80);

Array.Resize(bojanglesUUIDHash, 16);
var newV5Uuid = new Guid(bojanglesUUIDHash);

// To perform a proper comparison, the stringified expected v5 UUID result needs to be
// converted to a byte array and the first 3 components converted to little-endian
// representation, which is how Guids are stored internally:
var expectedResultBytes = new Guid("de9425a4-e8dd-510b-8e00-b6ac890c733a").ToByteArray();
Array.Reverse(expectedResultBytes, 0, 4);
Array.Reverse(expectedResultBytes, 4, 2);
Array.Reverse(expectedResultBytes, 6, 2);
var expectedResult = new Guid(expectedResultBytes);

// And now compare the results
Console.WriteLine(newV5Uuid.Equals(expectedResult));
// Should output 'false'.
```

To get an equivalent GUID using the same namespace GUID and 'bojangles' text, you could do either string manipulation (to convert the string to a byte array) or byte array manipulation on the namespace UUID so that the first 3 components have a Little Endian byte order. I'll do byte array manipulation since that's easier for me to do here:

```csharp
var g1 = new Guid("11de2b26-984e-56b4-aa25-b3bd28ea5ac2");
var namespaceBytes = g1.ToByteArray();
// Convert the bytes representing the a, b, and c components to little-endian byte order:
Array.Reverse(namespaceBytes, 0, 4);
Array.Reverse(namespaceBytes, 4, 2);
Array.Reverse(namespaceBytes, 6, 2);

var bojanglesBytes = System.Text.Encoding.UTF8.GetBytes("bojangles");

var bytes = new byte[namespaceBytes.Length + bojanglesBytes.Length];

Array.Copy(namespaceBytes, bytes, namespaceBytes.Length);
Array.Copy(bojanglesBytes, 0, bytes, namespaceBytes.Length, bojanglesBytes.Length);

var sha1Hasher = System.Security.Cryptography.SHA1CryptoServiceProvider.Create();
var bojanglesUUIDHash = sha1Hasher.ComputHash(bytes);

// Make sure this is a RFC4122 compliant v5 UUID...
bojanglesUUIDHash[6] = ((bojanglesUUIDHash[6] & 0x0F) | 0x50);
bojanglesUUIDHash[8] = ((bojanglesUUIDHash[8] & 0x3f) | 0x80);

Array.Resize(bojanglesUUIDHash, 16);
var newV5Uuid = new Guid(bojanglesUUIDHash);

// Again, to perform a proper comparison, the stringified expected v5 UUID needs some of its
// bytes reordered to Little-Endian ordering
var expectedResultBytes = new Guid("de9425a4-e8dd-510b-8e00-b6ac890c733a").ToByteArray();
Array.Reverse(expectedResultBytes, 0, 4);
Array.Reverse(expectedResultBytes, 4, 2);
Array.Reverse(expectedResultBytes, 6, 2);
var expectedResult = new Guid(expectedResultBytes);

Console.WriteLine(newV5Uuid..Equals(expectedResult));
// Should output 'true'
```

Now, it's true that if you `ToString` the `newV5Uuid` Guid, it won't output the same as according to an RFC4122-compliant UUID, as explained at the beginning. No Little-Endian to Big-Endian conversion is performed when stringifying a Guid:

```csharp
Console.WriteLine(newV5Uuid.ToString() == "de9425a4-e8dd-510b-8e00-b6ac890c733a");
// false
```

But, if we take the raw bytes in `newV5Uuid`, we'll see that the pure order of the bytes is exactly the same as the RFC4122-compliant UUID string representation:

```csharp
var sb = new StringBuilder();
var bytes = newV5Uuid.ToByteArray();
for (int i = 0; i < bytes.Length; ++i)
{
    sb.Append(bytes[i].ToString("x2"))
    if (i == 3 || i == 5 || i == 7 || i == 9)
    {
        sb.Append("-")
    }
}

Console.WriteLine(sb.ToString() == "de9425a4-e8dd-510b-8e00-b6ac890c733a");
// Should output 'true'
```

The above proves that the two unique identifiers are in fact identical, even if their string representations are different. Admittedly, what `System.Guid` really needs is an extension method, `.ToRfc4122String()`.

**Comment by abock** [comment]
Thanks for the feedback - I appreciate the time and effort you put into describing the issue. I hope I've addressed your ask in the latest commit/release.

I am very aware of the internals of `System.Guid`, and I do not think this is a real problem in .NET. UUID/GUIDs round trip just fine as-is. Further, .NET is fully RFC 4122 compliant. The specification itself allows for a mixed (LE,LE,LE,BE,BE) and fully BE encodings. BE is by far the most common encoding.

However, there is certainly a case to be made for [outputting the first three components in LE](https://en.wikipedia.org/wiki/Universally_unique_identifier#Encoding). As such I have added a `--mixed-encoding` (`-me` for shorthand) option for all of the GUID generators. This is available the 2.0 version that I will publish to NuGet shortly.

```bash
$ idgen v5 bojangles 11de2b26-984e-56b4-aa25-b3bd28ea5ac2
de9425a4-e8dd-510b-8e00-b6ac890c733a

$ idgen v5 --mixed-encoding bojangles 11de2b26-984e-56b4-aa25-b3bd28ea5ac2
a42594de-dde8-0b51-8e00-b6ac890c733a
```

**Comment by fourpastmidnight** [comment]
Thanks! Yes, it may have seemed I implied.Net was not RFC4122 compliant—but I know it is. I was trying to impress that the string output of .NET Guids are not round-trippable between, say, .Net and Java (and probably JavaScript).

So you’re addition it’s very welcome! Thanks again. I look forward to updating.

Sent from my Windows 10 phone
